File: backend/app/core/ids.py
"""Public and internal identifier helpers. Matches the TypeScript domain."""

from __future__ import annotations

import re
import uuid
from datetime import datetime, timezone


def new_id(prefix: str) -> str:
    return f"{prefix}_{uuid.uuid4()}"


def request_id() -> str:
    return f"req_{uuid.uuid4().hex[:20]}"


def format_public_id(kind: str, n: int, year: int | None = None) -> str:
    year = year if year is not None else datetime.now(timezone.utc).year
    if kind == "lead":
        return f"LD-{year}-{str(n).zfill(6)}"
    if kind == "delivery":
        return f"DLV-{year}-{str(n).zfill(6)}"
    return f"RULE-{str(n).zfill(3)}"


def public_id_number(public_id: str) -> int:
    match = re.search(r"(\d+)$", public_id)
    return int(match.group(1)) if match else 1


def counter_start(kind: str) -> int:
    if kind == "lead":
        return 1000
    if kind == "delivery":
        return 4800
    return 10

File: showcase/src/lib/pipeline/ids.ts
export function newId(prefix: string): string {
  const uuid =
    typeof crypto !== "undefined" && crypto.randomUUID
      ? crypto.randomUUID()
      : `${Date.now().toString(16)}-${Math.random().toString(16).slice(2)}`;
  return `${prefix}_${uuid}`;
}

export function requestId(): string {
  const uuid =
    typeof crypto !== "undefined" && crypto.randomUUID
      ? crypto.randomUUID().replace(/-/g, "").slice(0, 20)
      : Math.random().toString(36).slice(2, 12);
  return `req_${uuid}`;
}

export function formatPublicId(kind: "lead" | "delivery" | "rule", n: number, year = 2026): string {
  if (kind === "lead") return `LD-${year}-${String(n).padStart(6, "0")}`;
  if (kind === "delivery") return `DLV-${year}-${String(n).padStart(6, "0")}`;
  return `RULE-${String(n).padStart(3, "0")}`;
}

export function publicIdNumber(publicId: string): number {
  const match = publicId.match(/(\d+)$/);
  return match ? Number(match[1]) : 1;
}

export function sha256Hex(value: string): Promise<string> {
  const encoded = new TextEncoder().encode(value);
  if (typeof crypto !== "undefined" && crypto.subtle) {
    return crypto.subtle.digest("SHA-256", encoded).then((buf) => {
      return Array.from(new Uint8Array(buf))
        .map((b) => b.toString(16).padStart(2, "0"))
        .join("");
    });
  }
  // Fallback used only in limited test environments.
  let hash = 0;
  for (let i = 0; i < value.length; i += 1) hash = (hash * 31 + value.charCodeAt(i)) >>> 0;
  return Promise.resolve(hash.toString(16).padStart(16, "0"));
}
